### DISCUSSION TITLE
Mesh factory and optimization

### DISCUSSION THREAD

**Discussion by JezSonic** [MaSzyna-Reloaded/MaSzyna-API-wrapper#30] [Ideas]
A big conceptual change.

Since Godot is slowing down with a few dozen thousand nodes, I have been looking for a solution to escape from using tree nodes while preserving the godot scene format (scn/tscn) for storing data.

It is possible to declare custom resources that are saved in files but do not create nodes in the scene tree. All large data sets will go to resources and not nodes during import. Only a few proxies are needed for editing (TrackEditNode, TractionEditNode, etc.), as well as for displaying (rendering). These will be node tracks, traction, events, sounds. If someone in the original incs would like to mark foliage sets/blocks (especially grass), that would also be a lot of benefits.

 This is quite a subversion of the previous approach and requires a lot of changes, but it gives two important benefits:
- faster scene loading / no problem with overloaded godot tree
- moving the track drawing to the core porting library from the original Maszyn exe.

 That is, a potential godotwy exe (game, k and editor will use one "core" addon (libmaszyna), in which we do some physics and it will be possible to add mesh fabrication (e.g. track to start with). The editor and exe (game) may have different implementations of track mesh "rendering" (chunks, cullings, baking), but visualization consistency will be maintained.

Therefore, another interesting topic for the wrapper will be porting mesh factories from the original TTrack and TSegment to godot meshes. I am already drawing one rail more or less (but still testing in gdscript). And since we are moving to sharing custom types, it seems that the editor will have to become dependent on `libmaszyna`, and we will start placing custom resources of the TrackSegment type in the lib.

**Comment by marcinn** [comment]
This is how it should be done. Thanks.

**Comment by marcinn** [comment]
This is finally implemented in multiple servers and will be merged soon into main
